modules/data-entry/src/main/frontend/src/questionnaireEditor/ReferenceInput.jsx

```diff
@@ -0,0 +1,292 @@
+    // It is a list of either strings (for options) or recursive lists
+    // Each recursive list must have a string for its 0th option, which
+    // is taken to be its title
+    let newFilterableFields = [""];
+    // newFilterableTitles is a mapping from a string in newFilterableFields to a human-readable title
+    let newFilterableTitles = {"": "None"};
+    // newFilterablePaths is a mapping from a string in newFilterableFields to a path in the JCR
+    let newFilterablePaths = {"": ""};
+
+    // We'll need a helper recursive function to copy over data from sections/questions
+    let parseSectionOrQuestionnaire = (sectionJson, path="") => {
+      let retFields = [];
+      for (let [title, object] of Object.entries(sectionJson)) {
+        // We only care about children that are cards:Questions or cards:Sections
+        if (object["jcr:primaryType"] == "cards:Question") {
+          // If we are only interested in numeric values, ignore this one unless it is numeric
+          if (isNumeric && !(object["dataType"] == 'decimal' || object["dataType"] == 'long')) {
+            continue;
+          }
+
+          // If this is an cards:Question, copy the entire thing over to our Json value
+          retFields.push(object["jcr:uuid"]);
+          newFilterableTitles[object["jcr:uuid"]] = object["text"];
+          newFilterablePaths[object["jcr:uuid"]] = object["@path"];
+        } else if (object["jcr:primaryType"] == "cards:Section") {
+          // If this is an cards:Section, recurse deeper
+          retFields.push(...parseSectionOrQuestionnaire(object, path+title+"/"));
+        }
+        // Otherwise, we don't care about this value
+      }
+
+      return retFields;
+    }
+
+    // From the questionnaire homepage, we're looking for children that are objects of type cards:Questionnaire
+    for (let [title, thisQuestionnaire] of Object.entries(questionnaireJson)) {
+      if (thisQuestionnaire["jcr:primaryType"] != "cards:Questionnaire") {
+        continue;
+      }
+
+      newFilterableFields.push([title, ...parseSectionOrQuestionnaire(thisQuestionnaire)]);
+    }
+
+    // We also need a filter over the subject
+    setOptions(newFilterableFields);
+    setTitleMap(newFilterableTitles);
+    setPathMap(newFilterablePaths);
+  }
+
+  let parseSubjectTypeData = (subjectTypeJson) => {
+    // Parse through, but keep a custom field for the subject
+    let fields = [""];
+    let titles = {"": "None"};
+    let paths = {"": ""};
+
+    for (let subjectType of subjectTypeJson["rows"]) {
+      // For each reference, store its UUID and title
+      fields.push(subjectType["jcr:uuid"]);
+      titles[subjectType["jcr:uuid"]] = subjectType["label"];
+      paths[subjectType["jcr:uuid"]] = subjectType["@path"];
+    }
+    setOptions(fields);
+    setTitleMap(titles);
+    setPathMap(paths);
+  }
+
+  // From an array of fields, turn it into a react component
+  let GetReactComponentFromFields = (fields, nested=false) => {
```

Comment:
This is also a plain function, not a react component, so it should start with a lowercase letter.

```suggestion
  let getReactComponentFromFields = (fields, nested=false) => {
```